niri-config/src/lib.rs

```diff
@@ -2370,6 +2370,65 @@ pub enum PreviewRender {
     ScreenCapture,
 }
 
+#[derive(Debug, Clone)]
+pub enum ConfigPath {
+    /// Load the config only from this path, never create it.
+    Explicit(PathBuf),
+    /// Normal operation: Prioritize the user path, fallback to the system path,
+    /// else maybe create the user path if we're starting (but not reloading).
+    Regular {
+        user_path: PathBuf,
+        system_path: PathBuf,
+    },
+}
+
+impl ConfigPath {
+    pub fn load(&self) -> miette::Result<Config> {
+        let _span = tracy_client::span!("ConfigPath::load");
+
+        self.load_inner(|user_path, system_path| {
+            Err(miette::miette!(
+                "No config file found. Create one at {user_path} or {system_path}",
+                user_path = user_path.display(),
+                system_path = system_path.display(),
+            ))
+        })
+        .context("error loading config")
+    }
+
+    // Those lifetimes are really not flexible at all, but it's all we need for niri.
+    pub fn load_or_create_with<'a>(
+        &'a self,
+        maybe_create: impl FnOnce(&'a Path, &'a Path) -> miette::Result<&'a Path>,
+    ) -> miette::Result<Config> {
+        let _span = tracy_client::span!("ConfigPath::load_or_create_with");
+
+        self.load_inner(maybe_create)
+            .context("error loading config")
```

Comment:
Maybe this `.context()` should be inside `load_inner()`?

Reply:
or actually I think in the original we would only have this context in the watcher and not in the initial loading? I think the intention was that if we print "error creating config" then we shouldn't also print "error loading config". What would be the right place to put it?

Reply:
I refactored a bunch of the creation logic. The error for "failed to create config" now looks like this:

```sh
$ XDG_CONFIG_HOME=/nonexistent-nonwritable/ niri
INFO niri: starting version 25.05.1 (v25.05.1-47-g0f7e7791-modified)
ERROR niri:   × error loading config
  ├─▶ error creating config at "/nonexistent-nonwritable/niri/config.kdl"
  ├─▶ error creating config directory "/nonexistent-nonwritable/niri"
  ╰─▶ Permission denied (os error 13)
  ```
  
also notably, `WARN` -> `ERROR` because this is in fact an error.

Reply:
> Maybe this `.context()` should be inside `load_inner()`?

It should not be. If it's placed at the end of that function (`fn load_inner(...) { Config::load_internal(...).context(...) }`), then we lose the `error loading config` context when there was an error in creating the config file. I think that context *should* stay as the top level, *even* if we never did actually try to "load" from any path.

Reply:
> also notably, `WARN` -> `ERROR` because this is in fact an error.

Our `error!()` is for programming errors and after considering this case again, I don't think it warrants an exception, so I changed it back